### PR TITLE
503: Caching JWKS data

### PR DESCRIPTION
**Pull request by dbadham-fr**
This PR creates a new module securebanking-openbanking-uk-gateway-jwks which provides jwks related extensions for IG. This module  supports fetching (and optionally caching) jwks data.

The JwkSetService interface is used to control the behaviour when fetching jwks, the default implementation: RestJwkSetService will always fetch data using a rest api call. The CachingJwkSetService provides caching support via a pluggable Cache interface, when a cache miss occurs it will delegate to the RestJwkSetService to fetch the data.

CaffeineCachingJwkSetService provides a concrete implementation which uses the [caffeine](https://github.com/ben-manes/caffeine) library (this lib is already used within IG).

Caching behaviour can be controlled via IG config.
A heap object with name: `OBJwkSetService` of type `com.forgerock.securebanking.uk.gateway.jwks.JwkSetService` is a dependency of routes which need to fetch jwks data. 

Example config snippet to enable caching:
```
{
      "name": "OBJwkSetService",
      "type": "com.forgerock.securebanking.uk.gateway.jwks.cache.caffeine.CaffeineCachingJwkSetService",
      "config": {
        "maxCacheEntries": 500,
        "expireAfterWriteDuration": "12 hours"
      }
}
```
Example config snippet with no caching:
```
{
      "name": "OBJwkSetService",
      "type": "com.forgerock.securebanking.uk.gateway.jwks.RestJwkSetService"
}
```

The existing ProcessDetachedSig route has been updated to use the caching functionality, which will speed up processing. A typical jwks fetch from the OB directory takes 100-200ms on average.
Routes which add the ProcessDetachedSig filter must add the new filter arg: `"jwkSetService": "${heap['OBJwkSetService']}"`

Clients can use an alternative caching library by writing an adaptor class which implements our [Cache](https://github.com/SecureBankingAccessToolkit/securebanking-openbanking-uk-gateway/pull/162/files#diff-9a577de33318ae40eee4c8292045b0f350837ff179439d1629e2855c1263a93e) interface. The contents of package `com.forgerock.securebanking.uk.gateway.jwks.cache.caffeine` demonstrate how this can be done, and how to create the necessary objects to install such an implementation via IG config (using a heaplet object)

Issue: https://github.com/SecureBankingAccessToolkit/SecureBankingAccessToolkit/issues/503

### CODE REVIEW

**Comment by dbadham-fr** [comment]
The pom.xml in the root dir of this repo has been nabbed from: https://github.com/SecureBankingAccessToolkit/securebanking-openbanking-uk-gateway/pull/161 , so has a portion of the change to pr.yml

Work is required to tidy up the poms and build process.

File: config/7.0/obdemo-bank/ig/config/prod/config/config.json [hunk 1]
```diff
@@ -189,6 +189,10 @@
           }
         ]
       }
+    },
+    {
+      "name": "OBJwkSetService",
+      "type": "com.forgerock.securebanking.uk.gateway.jwks.cache.caffeine.CaffeineCachingJwkSetService"
     }
```
Comment (jorgesanchezperez) [inline config/7.0/obdemo-bank/ig/config/prod/config/config.json:196]: Should we decide if we want to set the cache in prod environment, don't?

**Review by jorgesanchezperez — COMMENTED** [review]
It seem good to me! nice job!

**Review by jorgesanchezperez — APPROVED** [review]

**Comment by BohoCode** [comment]
I agree, this looks great. Will speed up the service no end!!

**Review by jorgesanchezperez — APPROVED** [review]

**Review by jorgesanchezperez — APPROVED** [review]

**Pull request merged**